File: resources/android/StepperControl.kt
package dev.firstlightui.plugins.firstlight_ui.ui

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.defaultMinSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.width
import androidx.compose.material3.IconButton
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.clearAndSetSemantics
import androidx.compose.ui.semantics.disabled
import androidx.compose.ui.semantics.error
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.semantics.stateDescription
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import com.nativephp.plugins.native_ui.NativeUITokens

@Composable
internal fun FirstlightStepperControl(
    state: StepperRendererState,
    tokens: NativeUITokens,
    onDecrement: () -> Unit,
    onIncrement: () -> Unit,
    modifier: Modifier = Modifier,
) {
    val configuration = state.configuration
    val supporting = configuration.error.ifEmpty { configuration.helper }
    val description = listOf(
        configuration.accessibilityLabel,
        configuration.accessibilityHint,
        supporting.takeIf { configuration.error.isEmpty() }.orEmpty(),
        "Error: ${configuration.error}".takeIf { configuration.error.isNotEmpty() }.orEmpty(),
    ).filter(String::isNotEmpty).joinToString(". ")

    Column(modifier = modifier.fillMaxWidth(), verticalArrangement = Arrangement.spacedBy(6.dp)) {
        if (configuration.label.isNotEmpty()) {
            Text(
                text = configuration.label,
                color = tokens.onSurface,
                fontWeight = FontWeight.Medium,
                modifier = Modifier.clearAndSetSemantics {},
            )
        }

        Row(
            modifier = Modifier
                .fillMaxWidth()
                .defaultMinSize(minHeight = 48.dp)
                .semantics(mergeDescendants = true) {
                    if (description.isNotEmpty()) contentDescription = description
                    stateDescription = configuration.displayValue
                    if (configuration.disabled || state.isAwaitingPublication) disabled()
                    if (configuration.error.isNotEmpty()) error(configuration.error)
                },
            verticalAlignment = Alignment.CenterVertically,
        ) {
            IconButton(
                onClick = onDecrement,
                modifier = Modifier.semantics {
                    contentDescription = "Decrease ${configuration.accessibilityLabel}"
                },
                enabled = configuration.canPressDecrement && !state.isAwaitingPublication,
            ) {
                Text(text = "−", color = tokens.onSurface)
            }

            Spacer(Modifier.width(8.dp))
            Text(
                text = configuration.displayValue,
                color = tokens.onSurface,
                modifier = Modifier.defaultMinSize(minWidth = 48.dp),
            )
            Spacer(Modifier.width(8.dp))

            IconButton(
                onClick = onIncrement,
                modifier = Modifier.semantics {
                    contentDescription = "Increase ${configuration.accessibilityLabel}"
                },
                enabled = configuration.canPressIncrement && !state.isAwaitingPublication,
            ) {
                Text(text = "+", color = tokens.onSurface)
            }
        }

        if (supporting.isNotEmpty()) {
            Text(
                text = supporting,
                color = if (configuration.error.isEmpty()) tokens.onSurfaceVariant else tokens.destructive,
                modifier = Modifier.clearAndSetSemantics {},
            )
        }
    }
}

File: resources/ios/StepperControl.swift
import SwiftUI

struct FirstlightStepperControl: View {
    @Binding var state: StepperRendererState
    let tokens: NativeUITokens
    let onEvent: @MainActor (StepperRendererEvent) -> Void

    private var configuration: StepperRendererConfiguration { state.configuration }

    var body: some View {
        VStack(alignment: .leading, spacing: 6) {
            Stepper(
                onIncrement: incrementAction,
                onDecrement: decrementAction
            ) {
                HStack(alignment: .firstTextBaseline, spacing: 12) {
                    if !configuration.label.isEmpty {
                        Text(configuration.label)
                            .foregroundStyle(tokens.onSurface)
                            .fixedSize(horizontal: false, vertical: true)
                    }

                    Spacer(minLength: 12)

                    Text(configuration.displayValue)
                        .font(.body.monospacedDigit())
                        .foregroundStyle(tokens.onSurface)
                        .accessibilityHidden(true)
                }
            }
            .disabled(configuration.disabled || state.isAwaitingPublication)
            .accessibilityLabel(Text(configuration.accessibilityLabel))
            .accessibilityHint(Text(accessibilityHint))
            .accessibilityValue(Text(configuration.displayValue))

            let supporting = configuration.error.isEmpty ? configuration.helper : configuration.error
            if !supporting.isEmpty {
                Text(supporting)
                    .font(.footnote)
                    .foregroundStyle(configuration.error.isEmpty ? tokens.onSurfaceVariant : tokens.destructive)
                    .fixedSize(horizontal: false, vertical: true)
                    .accessibilityLabel(configuration.error.isEmpty ? supporting : "Error: \(supporting)")
            }
        }
        .opacity(configuration.disabled ? 0.6 : 1)
    }

    private var decrementAction: (() -> Void)? {
        guard configuration.canPressDecrement else { return nil }
        return {
            if let event = state.decrement() { onEvent(event) }
        }
    }

    private var incrementAction: (() -> Void)? {
        guard configuration.canPressIncrement else { return nil }
        return {
            if let event = state.increment() { onEvent(event) }
        }
    }

    private var accessibilityHint: String {
        [
            configuration.accessibilityHint,
            configuration.error.isEmpty ? configuration.helper : "Error: \(configuration.error)",
        ]
        .filter { !$0.isEmpty }
        .joined(separator: ". ")
    }
}
